testdata/generate_test_data/create_complex_circular_references.py

```diff
@@ -0,0 +1,203 @@
+
+
+def create_circular_references_test_graph(replication_counter: int) -> etree._Element:
+    """
+    This function creates a graph with circular references.
+    It is capable of reproducing one graph with all the references a specified number of times.
+    Each circle type has its own sub-graph number in order to differentiate them by their ID.
+    The resulting etree is suitable for an upload to the DSP-API.
+
+    Args:
+        replication_counter: number of times the sub-graphs should be created in one root-graph.
+
+    Returns:
+        A etree which is suitable for an upload into the DSP-API
+    """
+    root = excel2xml.make_root("0700", "simcir")
+    for i in range(1, replication_counter + 1):
+        root.extend(_make_one_circle_with_three_resources(replication_counter=f"{i}1"))
+        root.extend(_make_complex_dependencies(replication_counter=f"{i}2"))
+        root.extend(_make_complex_dependencies_add_on(replication_counter=f"{i}3"))
+        root.extend(_make_two_references(replication_counter=f"{i}4"))
+        root.extend(_make_chain(replication_counter=f"{i}5"))
+        root.extend(_make_inverted_complex_dependencies(replication_counter=f"{i}6"))
+    return root
+
+
+def _make_list_of_resources(number_of_resources: int, replication_counter: str, start_range: int = 65) -> list[Any]:
+    letters = [chr(x) for x in range(start_range, start_range + number_of_resources)]
+    return [
+        excel2xml.make_resource(
+            restype=":TestThing", label=f"res_{let}_{replication_counter}", id=f"res_{let}_{replication_counter}"
+        )
+        for let in letters
+    ]
+
+
+def _make_salsah_link(target_res: etree._Element) -> str:
+    return f'<a class="salsah-link" href="IRI:{target_res.attrib["id"]}:IRI">{target_res.attrib["id"]}</a>'
+
+
+def _make_xml_text_prop(target_res: etree._Element | list[etree._Element]) -> etree._Element:
+    match target_res:
+        case etree._Element():
+            salsah_link = _make_salsah_link(target_res)
+            # one resource with many targets
+        case list():
+            salsah_link = "".join([_make_salsah_link(x) for x in target_res])
+    return excel2xml.make_text_prop(name=":hasRichtext", value=excel2xml.PropertyElement(salsah_link, encoding="xml"))
+
+
+def _make_resptr_prop(target_res: list[etree._Element] | etree._Element) -> etree._Element:
+    match target_res:
+        case etree._Element():
+            link = excel2xml.make_resptr_prop(name=":hasResource", value=target_res.attrib["id"])
+        # one resource with many targets
+        case list():
+            link = excel2xml.make_resptr_prop(name=":hasResource", value=[x.attrib["id"] for x in target_res])
+    return link
+
+
+def _make_simple_text(resource: etree._Element) -> etree._Element:
+    resource.append(excel2xml.make_text_prop(name=":hasSimpleText", value="This is text."))
+    return resource
```

Comment:
if you do an in-place modification, there is really no need to return the object, because every caller already has a reference to the modified object

Reply:
My problem here is, that if I do this in a list comprehension, I get a list back from the function which contains only None values, because the function does not return anything.

Reply:
I always considered it good practice to return modified objects, so that for the caller, it is easier to know that the function modified that object. Imagine this case:

```
my_object = create_object()
function_with_bad_name()  # silently modifies my_object
# my_object is now modified, but I don't know!
```

I already dislike that many functions modify the objects they receive, but then I try to always return them, to make things clearer. 

What do you think?